pkg/acpi/raw.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package acpi
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/u-root/u-root/pkg/io"
+)
+
+// Raw ACPI table support. Raw tables are those tables
+// one needs to read in, write out, but not change in any way.
+// This is needed when, e.g., a program has to reassemble all the
+// tables in /sys for kexec.
+
+// Raw is just a table embedded in a []byte.  Operations on Raw are
+// useful for unpacking into a more refined table or just figuring out
+// how to skip a table you don't care about.
+type Raw struct {
+	data []byte
+}
+
+var _ = Tabler(&Raw{})
+
+// NewRaw returns a new Raw table given a byte slice.
+func NewRaw(b []byte) (Tabler, error) {
+	u := binary.LittleEndian.Uint32(b[LengthOffset : LengthOffset+4])
+	return &Raw{data: b[0:u]}, nil
+}
+
+// RawFromFile reads a raw table in from a file.
+func RawFromFile(n string) (Tabler, error) {
+	b, err := ioutil.ReadFile(n)
+	if err != nil {
+		return nil, err
+	}
+	return NewRaw(b)
+}
+
+// ReadRaw reads a full table in, given an address.
+// ReadRaw uses the io package. This may not always work
+// if the kernel has restrictions on reading memory above
+// the 1M boundary, and the tables are above boundary.
+func ReadRaw(a int64) (Tabler, error) {
+	var u io.Uint32
+	// Read the table size at a+4
+	if err := io.Read(a+4, &u); err != nil {
+		return nil, err
+	}
+	Debug("ReadRaw: Size is %d", u)
+	dat := make([]byte, u)
+	for i := range dat {
+		var d io.Uint8
+		if err := io.Read(a+int64(i), &d); err != nil {
```

Comment:
This opens "/dev/mem" every time you call it. Hopefully it's not slow.

Reply:
we've decided it's not "too slow", but yes, it's slow.